UIforETW/Utility.cpp

```diff
@@ -20,7 +20,305 @@ limitations under the License.
+		return false;
+	}
+	(*hmem) = mem;
+	return true;
+}
+
+void FreeGlobalMemory(_Frees_ptr_opt_ const HGLOBAL hmem)
+{
+		//If [GlobalFree] succeeds, the return value is NULL.
+		const HGLOBAL freeRes = ::GlobalFree(hmem);
+		if (freeRes != NULL)
+			std::terminate();//Logic bug!
+}
+
+_Success_(return)
+bool LockGlobalMemory(_In_ _Pre_valid_ const HGLOBAL hmem, _Outptr_ PVOID* const ptrMem)
+{
+	void* const ptr = ::GlobalLock(hmem);
+	if (ptr == NULL)
+	{
+		outputPrintf(L"Failed to lock memory for clipboard text!\n");
+		outputLastError();
+		return false;
+	}
+	(*ptrMem) = ptr;
+	return true;
+}
+
+void UnlockGlobalMemory(_In_ const HGLOBAL hmem)
+{
+	const BOOL unlockRes = ::GlobalUnlock(hmem);
+	if (!unlockRes)
+	{
+		const DWORD lastErr = ::GetLastError();
+		if (lastErr != NO_ERROR)
+			std::terminate( );//Logic bug!
+	}
+}
+
+_Success_(return)
+bool SizeGlobalMemory(_In_ _Pre_valid_ const HGLOBAL hmem, _Out_ size_t* const sizeBytes)
+{
+	const size_t bytes = ::GlobalSize(hmem);
+	if (bytes == 0)
+	{
+		outputPrintf(L"Failed to get size of globally-allocated memory!\n");
+		outputLastError();
+		return false;
+	}
+	(*sizeBytes) = bytes;
+	return true;
+}
+
+_Success_(return)
+bool ClipboardGetUnicode(_Out_ _Post_valid_ HANDLE* const hClipboard)
+{
+	const HANDLE hClip = ::GetClipboardData(CF_UNICODETEXT);
+	if (hClip == NULL)
+	{
+		outputPrintf(L"Failed to get clipboard data!\n");
+		outputLastError();
+		return false;
+	}
+	(*hClipboard) = hClip;
+	return true;
+}
+
+_Success_(return)
+bool ClipboardSetUnicode(_In_ const HGLOBAL hmem)
+{
+	const HANDLE dataHandle = ::SetClipboardData(CF_UNICODETEXT, hmem);
+	if (dataHandle == NULL)
+	{
+		outputPrintf(L"Failed to set clipboard data!\n");
+		outputLastError();
+		return false;
+	}
+	return true;
+}
+
+void closeClipboard()
+{
+	const BOOL closeResult = ::CloseClipboard( );
+	if (closeResult)
+		return;
+	debugPrintf(L"Failed to close the clipboard!\n");
+	debugLastError( );
+}
+
+bool openClipboard()
+{
+	const BOOL openClip = ::OpenClipboard(::GetDesktopWindow());
+	if (!openClip)
+	{
+		outputPrintf(L"Failed to open clipboard!\n");
+		debugLastError();
+		return false;
+	}
+	return true;
+}
+
+bool emptyClipboard()
+{
+	const BOOL emptyClip = ::EmptyClipboard();
+	if (!emptyClip)
+	{
+		outputPrintf(L"Failed to empty clipboard!\n");
+		debugLastError();
+		return false;
+	}
+	return true;
+}
+
+_Success_(return)
+bool OpenRegKey( _Out_ HKEY* const key, _In_ const HKEY root, PCWSTR const subkey )
+{
+	const LONG openResult = ::RegOpenKeyExW(root, subkey, 0, KEY_ALL_ACCESS, key);
+	if (openResult == ERROR_SUCCESS)
+		return true;
+	outputPrintf(L"Failed to open registry key `%s`.\n", subkey);
+	debugLastError();
+	return false;
+}
+
+void CloseRegKey(_In_ _Pre_valid_ _Post_ptr_invalid_ const HKEY key, PCWSTR const keyName)
+{
+	const LONG closeKey = ::RegCloseKey(key);
+	if (closeKey == ERROR_SUCCESS)
+		return;
+	outputPrintf(L"Failed to close registry key `%s`.\n", keyName);

```

Comment:
I can't actually remember why I decided to *log* this instead of `std::terminate()`ing, or just ignoring it. A failed `RegCloseKey` either means:

1. A logic error (like `CloseHandle`)
2. We're leaking a handle, and may do so several times:
![uiforetw_regclosekey](https://cloud.githubusercontent.com/assets/2142308/10557485/d63d1a40-747c-11e5-8f85-1550bf860b08.PNG)

...it may be totally ok to leak a few registry key handles, but I prefer not to. Should I just remove the check?